### ISSUE TITLE
slicing tensors within the network

### ISSUE THREAD

**Issue by aosokin** [pytorch/pytorch#822]
Hi, I've recently observed something weird which might indicate an obscure bug.
Very similar issue was open [here](https://github.com/hughperkins/pytorch/issues/20).

I'm defining a network, and in forward() I have a slicing operation separating the first half of my channels:
`
first_half = input[:, :self.n_input_ch // 2, :, :]
`
This operation totally breaks down the training, i.e., after two or three training iterations the loss goes to nan. On top of that, something weird happens with variable references meaning some unrelated tensors start to randomly change.

However, if I do
`
first_half = input[:, :self.n_input_ch // 2, :, :] * 1.0
`
instead, then everything works fine.

Does this look like  a bug or slicing is not supposed to be used this way?

**Comment by apaszke** [comment]
It is supposed to be used this way. You're likely doing some in place modifications to sliced tensors, that for some reason aren't detected. Can you please try to write a small script that can reproduce the problem?

**Comment by aosokin** [comment]
Hi again,
I've started from [the default MNIST example](https://github.com/pytorch/examples/blob/master/mnist/main.py) and have changed the network to have two streams.
Two versions:
1. have two Conv2d layers and just torch.cat features coming from them - runs, but produces bad accuracy, although mathematically is identical to the default network. I suspect that the gradient computation is wrong.
2 .slice the features in two halves, apply to Conv2d layers, torch.cat the result - crashes after processing several batches.

Here is the code:

`from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=10, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='enables CUDA training')
parser.add_argument('--seed', type=int, default=1, metavar='S',
                    help='random seed (default: 1)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
args = parser.parse_args()
args.cuda = not args.no_cuda and torch.cuda.is_available()

torch.manual_seed(args.seed)
if args.cuda:
    torch.cuda.manual_seed(args.seed)


kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
    ])),
    batch_size=args.batch_size, shuffle=True, **kwargs)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)

        # standard
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)

        # CHANGE v1: no slicing, only concat in the network
        # Produces very bad results, probably the gradient is wrong
        self.conv2_1_v1 = nn.Conv2d(10, 10, kernel_size=5)
        self.conv2_2_v1 = nn.Conv2d(10, 10, kernel_size=5)

        # CHANGE v2: slicing + concat - crashes
        self.conv2_1_v2 = nn.Conv2d(5, 10, kernel_size=5)
        self.conv2_2_v2 = nn.Conv2d(5, 10, kernel_size=5)

        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))

        # # standard
        x_std = self.conv2(x)
        # x = x_std

        # # CHANGE v1: should be identical to the standard version
        # x_1 = self.conv2_1_v1(x)
        # x_2 = self.conv2_2_v1(x)
        # x = torch.cat((x_1, x_2), 1)

        # CHANGE v2: two path
        x_1 = x[:, :5, :, :]
        x_2 = x[:, 5:, :, :]
        x_1 = self.conv2_1_v2(x_1)
        x_2 = self.conv2_2_v2(x_2)
        x = torch.cat((x_1, x_2), 1)

        print('Target size:', x_std.size())
        print('Obtained size:', x.size())

        x = F.relu(F.max_pool2d(self.conv2_drop(x), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = F.relu(self.fc2(x))
        return F.log_softmax(x)


model = Net()
if args.cuda:
    model.cuda()

optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)


def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        if args.cuda:
            data, target = data.cuda(), target.cuda()
        data, target = Variable(data), Variable(target)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.data[0]))


def test(epoch):
    model.eval()
    test_loss = 0
    correct = 0
    for data, target in test_loader:
        if args.cuda:
            data, target = data.cuda(), target.cuda()
        data, target = Variable(data, volatile=True), Variable(target)
        output = model(data)
        test_loss += F.nll_loss(output, target).data[0]
        pred = output.data.max(1)[1]  # get the index of the max log-probability
        correct += pred.eq(target.data).cpu().sum()

    test_loss = test_loss
    test_loss /= len(test_loader)  # loss function already averages over batch size
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


for epoch in range(1, args.epochs + 1):
    train(epoch)
    test(epoch)
`

**Comment by soumith** [comment]
@aosokin just for completeness, what pytorch version are you running? we fixed some bugs related to non-contiguous gradients in v0.1.9

**Comment by aosokin** [comment]
I'm running binaries released in conda, which is v0.1.8.
Will try v0.1.9 right now.

**Comment by aosokin** [comment]
@soumith, it looks the bug I was hitting is already fixed.
Now everything works. Thanks a lot for the quick reaction!

**Comment by lkqnaruto** [comment]
> @aosokin just for completeness, what pytorch version are you running? we fixed some bugs related to non-contiguous gradients in v0.1.9

Hi,

Just want to make sure, we can slice the input of the loss function?

For example:
y = model(input).      # y is a n by n matrix
loss_output = loss(y[2/n,:])

Thanks in advance
